ogusa/aggregates.py

```diff
@@ -242,40 +257,36 @@ def revenue(r, w, b, n, BQ, Y, L, K, factor, params):
     Returns: T_H
 
     '''
-
-    e, lambdas, omega, method, etr_params, tax_func_type, theta, tau_bq, \
-        tau_payroll, h_wealth, p_wealth, m_wealth, retire, T, S, J,\
-        tau_b, delta_tau = params
-
-    I = r * b + w * e * n
-
-    if I.ndim == 2:
-        T_I = np.zeros((S,J))
-        for j in range(J):
-            TI_params = (e[:,j], etr_params, tax_func_type)
-            T_I[:,j] = tax.ETR_income(r, w, b[:,j], n[:,j], factor, TI_params) * I[:,j]
-    if I.ndim == 3:
-        T_I = np.zeros((T,S,J))
-        for j in range(J):
-            if etr_params.ndim == 3:
-                tau_inc_params3D = etr_params[:,j,:]
-            if etr_params.ndim == 4:
-                tau_inc_params3D = etr_params[:,:,j,:]
-            TI_params = (e[:,:,j], tau_inc_params3D, tax_func_type)
-            T_I[:,:,j] = tax.ETR_income(r[:,:,j], w[:,:,j], b[:,:,j], n[:,:,j], factor, TI_params) * I[:,:,j]
-    T_P = tau_payroll * w * e * n
-    TW_params = (h_wealth, p_wealth, m_wealth)
-    T_W = tax.ETR_wealth(b, TW_params) * b
     if method == 'SS':
-        T_P[retire:] -= theta * w
-        T_BQ = tau_bq * BQ / lambdas
-        biz_params = (tau_b, delta_tau)
-        business_revenue = tax.get_biz_tax(w, Y, L, K, biz_params)
-        REVENUE = (omega * lambdas * (T_I + T_P + T_BQ + T_W)).sum() + business_revenue
+        I = r * b + w * p.e * n
+        T_I = np.zeros_like(I)
+        for j in range(p.J):
+            TI_params = (p.e[:, j], p.etr_params[-1:, :, :], p.tax_func_type) ## this is written for SS - check if ever go here in TPI
+            T_I[:, j] = tax.ETR_income(r, w, b[:, j], n[:, j], factor, TI_params) * I[:, j]
+        T_P = p.tau_payroll * w * p.e * n
+        T_P[p.retire:] -= theta * w
+        T_W = tax.ETR_wealth(b, p) * b
+        T_BQ = p.tau_bq * (BQ / np.transpose(p.lambdas))
+        business_revenue = tax.get_biz_tax(w, Y, L, K, p)
+        REVENUE = (np.transpose(p.omega_SS * p.lambdas) * (T_I + T_P + T_BQ + T_W)).sum() + business_revenue
     elif method == 'TPI':
-        T_P[:, retire:, :] -= theta.reshape(1, 1, J) * w[:,retire:,:]
-        T_BQ = tau_bq.reshape(1, 1, J) * BQ / lambdas
-        biz_params = (tau_b, delta_tau)
-        business_revenue = tax.get_biz_tax(w[:T,0,0], Y, L, K, biz_params)
-        REVENUE = (omega * lambdas * (T_I + T_P + T_BQ + T_W)).sum(1).sum(1) + business_revenue
+        I = ((np.tile(np.reshape(r, (p.T, 1, 1)), (1, p.S, p.J)) * b) +
+             (np.tile(np.reshape(w, (p.T, 1, 1)), (1, p.S, p.J)) *
+              (n * p.e)))
+        T_I = np.zeros_like(I)
+        for j in range(p.J):
+            TI_params = (p.e[:, j], p.etr_params, p.tax_func_type)
```

Comment:
@hdoupe Is there a better way to do this array broadcasting and avoiding all the `reshape` and `tile` commands?  With the refactoring going on in this PR, I'm trying to use just the parameter class rather than passing tuples of parameters to each function call.  However, that means the parameter arrays are the same size for all calculations (e.g. for the steady state and time path functions) and it results in lots of reshaping of things inside the functions.

Reply:
Oh, I see. It's pretty hard to read the nested `np.tile` and `np.reshape` calls. What are they doing? As in what is the shape of the array before the call and after the call?

Reply:
Do you think you could classify the different types of reshape/tile/squeeze calls? For example, could the sequence:
`np.tile(np.reshape(some_array, (p.T, 1, 1)), (1, p.S, p.J))` be a method like `to_timepath_shape` or `to_steadystate_shape` or similar?

Reply:
@hdoupe I like that suggestion - wouldn't have thought of it.  Thanks!

Reply:
Great, I'm happy to discuss other solutions or implementation details of this one.